kernel/comps/softirq/src/lock.rs

```diff
@@ -0,0 +1,179 @@
+// SPDX-License-Identifier: MPL-2.0
+
+use ostd::{
+    cpu_local_cell,
+    sync::{GuardTransfer, SpinGuardian},
+    task::{
+        atomic_mode::{AsAtomicModeGuard, InAtomicMode},
+        disable_preempt, DisabledPreemptGuard,
+    },
+    trap::{disable_local, in_interrupt_context},
+};
+
+use crate::process_all_pending;
+
+cpu_local_cell! {
+    static DISABLE_SOFTIRQ_COUNT: u8 = 0;
+}
+
+/// Returns whether softirq is enabled on local CPU.
+pub(super) fn is_softirq_enabled() -> bool {
+    DISABLE_SOFTIRQ_COUNT.load() == 0
+}
+
+/// A guardian that disables bottom half while holding a lock.
+pub enum BottomHalfDisabled {}
+
+/// A guard for disabled local softirqs.
+pub struct DisableLocalBottomHalfGuard {
+    preempt: DisabledPreemptGuard,
+}
+
+impl AsAtomicModeGuard for DisableLocalBottomHalfGuard {
+    fn as_atomic_mode_guard(&self) -> &dyn InAtomicMode {
+        &self.preempt
+    }
+}
+
+impl Drop for DisableLocalBottomHalfGuard {
+    fn drop(&mut self) {
+        // Once the guard is dropped, we will process pending items within
+        // the current thread's context if softirq is going to be enabled.
+        // This behavior is similar to how Linux handles pending softirqs.
+        if DISABLE_SOFTIRQ_COUNT.load() == 1 && !in_interrupt_context() {
+            // Preemption and softirq are not really enabled at the moment,
+            // so we can guarantee that we'll process any pending softirqs for the current CPU.
+            let irq_guard = disable_local();
+            let irq_guard = process_all_pending(irq_guard);
+
+            // To avoid race conditions, we should decrease the softirq count first,
+            // then drop the IRQ guard.
+            DISABLE_SOFTIRQ_COUNT.sub_assign(1);
+            drop(irq_guard);
+            return;
+        }
+
+        DISABLE_SOFTIRQ_COUNT.sub_assign(1);
+    }
+}
+
+#[must_use]
+fn disable_local_bottom_half() -> DisableLocalBottomHalfGuard {
+    // When disabling softirq, we must also disable preemption
+    // to avoid the task to be scheduled to other CPUs.
+    let preempt = disable_preempt();
+    DISABLE_SOFTIRQ_COUNT.add_assign(1);
+    DisableLocalBottomHalfGuard { preempt }
+}
+
+impl GuardTransfer for DisableLocalBottomHalfGuard {
+    fn transfer_to(&mut self) -> Self {
+        DISABLE_SOFTIRQ_COUNT.add_assign(1);
+        Self {
+            preempt: disable_preempt(),
+        }
```

Comment:
https://github.com/asterinas/asterinas/pull/1623#discussion_r2040936879

The current version is correct, but I'd like to know why you don't call `disable_local_bottom_half` directly?

Reply:
I didn't initially realize that you were referring to using the disable_local_bottom_half API directly. As a result, I only addressed the original issue of the softirq count not increasing.

Now, I use the suggested API.